File: src/main/java/org/bk/graphql/service/book/BookCommands.java
package org.bk.graphql.service.book;

import org.bk.graphql.domain.AuthorId;

import java.util.Set;

public interface BookCommands {
    record CreateBookCommand(String name, int pageCount, Set<AuthorId> authors) {
    }
    record CreateOrUpdateBookCommand(String id, String name, int pageCount, Set<AuthorId> authors, int version) {
        public CreateOrUpdateBookCommand(String id, String name, int pageCount, Set<AuthorId> authors) {
            this(id, name, pageCount, authors, 0);
        }
    }
    record UpdateBookCommand(String id, String name, int pageCount, Set<AuthorId> authors, int version) {
    }
    record UpdateBookNameCommand(String id, String name, int version) {
    }
}

File: src/main/kotlin/org/bk/graphql/service/BookCommands.kt
package org.bk.graphql.service

import org.bk.graphql.domain.AuthorId

data class CreateBookCommand(val name: String, val pageCount: Int, val authors: Set<AuthorId>)
data class CreateOrUpdateBookCommand(val id: String, val name: String, val pageCount: Int, val authors: Set<AuthorId>, val version: Int = 0)
data class UpdateBookCommand(val id: String, val name: String, val pageCount: Int, val authors: Set<AuthorId>, val version: Int)